url270fc4f5ee_ykcilborw_Joust_tgz-pJ8-fri_patterns_interpreter_parsergenerator_examples_HowToOverrideLexerImplJ8/src/fri/patterns/interpreter/parsergenerator/examples/Joust.java

```diff
@@ -0,0 +1,419 @@
+		{ "TYPE", "'P'"},
+		{ "TYPE", "'K'"},
+		{ "TYPE", "'Q'"},
+		{ "TYPE", "'N'"},
+		{ "TYPE", "'R'"},
+		{ "TYPE", "'B'"},
+		{ "TYPE", "'G'"},
+		{ "TYPE", "'g'"},
+		{ "TYPE", "'L'"},
+		{ "TYPE", "'l'"},
+		{ "TYPE", "'D'"},
+		{ "TYPE", "'d'"},
+		
+		// generic pieces
+		{ "TYPE", "'p''"},
+		{ "TYPE", "'k''"},
+		{ "TYPE", "'q''"},
+		{ "TYPE", "'n''"},
+		{ "TYPE", "'r''"},
+		{ "TYPE", "'b''"},
+		{ "TYPE", "'P''"},
+		{ "TYPE", "'K''"},
+		{ "TYPE", "'Q''"},
+		{ "TYPE", "'N''"},
+		{ "TYPE", "'R''"},
+		{ "TYPE", "'B''"},
+		
+		{ "X", "'a'"},
+		{ "X", "'b'"},
+		{ "X", "'c'"},
+		{ "X", "'d'"},
+		{ "X", "'e'"},
+		{ "X", "'f'"},
+		{ "X", "'g'"},
+		{ "X", "'h'"},
+		
+		{ "OPERATOR", "'&&'" },
+		{ "OPERATOR", "'||'" },
+		{ "OPERATOR", "'and'" },
+		{ "OPERATOR", "'or'" },
+		
+		{ "COMPARATOR", "'<'"},
+		{ "COMPARATOR", "'='"},
+		{ "COMPARATOR", "'>'"},
+		{ "COMPARATOR", "'>='"},
+		{ "COMPARATOR", "'<='"},
+		
+		{ "NOT", "'not'" },
+		{ "NOT", "'occurs'"},
+		
+		{ "KING", "'K'"},
+		{ "KING", "'k'"},
+		
+		{ "FACTOR",   "`number`", },
+		{ "VARIABLE",   "`letter`", },
+		{ "VARIABLE",   "VARIABLE", "`letter`", },
+		//{ "PROGRAMDELIMITER", "','" },
+		{ "PROGRAMDELIMITER", "'through'"},
+		{ "PROGRAMDELIMITER", "'until'"},
+		
+		
+		{ Token.IGNORED,   "`whitespaces`" },
+	};
+	
+	public Object PROGRAM (Object STATEMENT) {
+		ArrayList<Object> toReturn = new ArrayList<Object>();
+		toReturn.add(STATEMENT);
+		return toReturn;
+	}
+	
+	public Object PROGRAM (Object PROGRAM, Object delimiter, Object STATEMENT) {
+		String s = (String) delimiter;
+		ArrayList<Object> toReturn = new ArrayList<Object>();
+		if (s.equals("through") || s.equals("until")) {
+			ArrayList<Object> program = (ArrayList<Object>) PROGRAM;
+			//System.out.println("orig prog: " + program);
+			int size = program.size();
+			Object last = program.get(size - 1);
+			program.remove(size - 1);
+			ArrayList<Object> toReturn2 = new ArrayList<Object>();
+			toReturn2.add(delimiter);
+			toReturn2.add(last);
+			toReturn2.add(STATEMENT);
+			program.add(toReturn2);
+			//System.out.println("prog: " + program);
+			toReturn = program;
+		} else {
+			ArrayList<Object> program = (ArrayList<Object>) PROGRAM;
+			program.add(STATEMENT);
+			toReturn = program;
+		}
+		return toReturn;
+	}
+	
+	public Object STATEMENT(Object EXPRESSION) {
+		//System.out.println("STAT EXPRESSION: " + EXPRESSION);
+		return EXPRESSION;
+	}
+
+	
+	public Object STATEMENT(Object STATEMENT, Object OPERATOR, Object EXPRESSION) {
+		//return "[" + OPERATOR + ", " + STATEMENT + ", " + EXPRESSION + "]";
+		//return "[" + OPERATOR + " " + STATEMENT + " " + EXPRESSION + "]";
+		//System.out.println("STATEMENT: " + STATEMENT);
+		//System.out.println("OPERATOR: " + OPERATOR);
+		//System.out.println("EXPRESSION: " + EXPRESSION);
+		ArrayList<Object> toReturn = new ArrayList<Object>();
+		toReturn.add(OPERATOR);
+		toReturn.add(STATEMENT);
+		toReturn.add(EXPRESSION);
+		return toReturn;
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

Similar code fragments were detected in the same file at the following lines: 184:188, 210:214.  
Refactoring can help improve code maintainability. Consider reducing duplicate code by extracting it into a separate method. You can then replace duplicated code with calls to this new method.